### ISSUE TITLE
Updates.py format

### ISSUE THREAD

**Issue by skaae** [Lasagne/Lasagne#58]
I'm trying to understand the updates.py input/output format. 
1. **loss**: Theano expression calculating the cost, e.g. squared or nll
2. **all_params**: list from nntools.layers.get_all_params(top_layer). This is a list of a list of Theano shared variables representing the parameters
3. **learning_rate**: float

The SGD is easy to understand:

```
all_grads = [theano.grad(loss, param) for param in all_params]
updates = []

for param_i, grad_i in zip(all_params, all_grads):
    updates.append((param_i, param_i - learning_rate * grad_i))
```

Get the gradient expression for all parameters and return a list which calculates the updates.
The ouput format is list of tuples with [0]: initial parameter [1] updated parameter

I dont understand momentum and the following methods:

```
all_grads = [theano.grad(loss, param) for param in all_params]
updates = []

for param_i, grad_i in zip(all_params, all_grads):
    mparam_i = theano.shared(np.zeros(param_i.get_value().shape, dtype=theano.config.floatX))
    v = momentum * mparam_i - weight_decay * learning_rate * param_i  - learning_rate * grad_i
    updates.append((mparam_i, v))
    updates.append((param_i, param_i + v))
```

mparam_i: a shared tensor of zeros? 

In the calcualtion of v, wouldn't it be sensible to scale by (1-momentum) in the second part of the eq:
 v = momentum \* mparam_i -  (1-momentum)*(weight_decay \* learning_rate \* param_i  - 
                                                                           learning_rate \* grad_i)

What is the 2 tuples appended to updates? The second is the update what is the first and how is i handled?
